### DISCUSSION TITLE
Connected peers

### DISCUSSION THREAD

**Discussion by jonathanlocke** [ep2p/kademlia-api#65] [General]
Is there a way to get the peers that are known by a given node?

Thanks!

Jon

**Comment by sepgh** [comment]
Hi

Internally a node can get all peers by first [getting all buckets from the routing table](https://github.com/ep2p/kademlia-api/blob/89553583d3a0c95ad37239283065f2b3dad32535/src/main/java/io/ep2p/kademlia/table/RoutingTable.java#L57C5-L57C26) and read all nodes from a bucket. Right now there is no good API for it, but its easy to add that. You can [get all node ids](https://github.com/ep2p/kademlia-api/blob/89553583d3a0c95ad37239283065f2b3dad32535/src/main/java/io/ep2p/kademlia/table/Bucket.java#L28C5-L28C26) and then get [`ExternalNode` objects](https://github.com/ep2p/kademlia-api/blob/main/src/main/java/io/ep2p/kademlia/node/external/ExternalNode.java) using [`getNode(I id)` method of the bucket](https://github.com/ep2p/kademlia-api/blob/89553583d3a0c95ad37239283065f2b3dad32535/src/main/java/io/ep2p/kademlia/table/Bucket.java#L27).



If you want to go a step further and let other nodes ask for such info, you'd have to create new [message types](https://github.com/ep2p/kademlia-api/blob/main/src/main/java/io/ep2p/kademlia/protocol/MessageType.java) (just a string name) that can request this, and then write a [handler](https://github.com/ep2p/kademlia-api/blob/main/src/main/java/io/ep2p/kademlia/protocol/handler/MessageHandler.java) for them (request and response handlers), and handle serialization as well. 

At last you should register the handlers on nodes.
